gotham/src/router/builder/mod.rs

```diff
@@ -267,6 +267,382 @@ where
     }
 }
 
+/// Implements the methods required for associating a number of routes with a single path. See
+/// `DrawRoutes::associated`.
+pub struct AssociatedRouteBuilder<'a, C, P, PE>
+where
+    C: PipelineHandleChain<P> + Copy + Send + Sync + 'static,
+    P: Send + Sync + 'static,
+    PE: PathExtractor + Send + Sync + 'static,
+{
+    node_builder: &'a mut NodeBuilder,
+    pipeline_chain: C,
+    pipelines: PipelineSet<P>,
+    phantom: PhantomData<PE>,
+}
+
+impl<'a, C, P, PE> AssociatedRouteBuilder<'a, C, P, PE>
+where
+    C: PipelineHandleChain<P> + Copy + Send + Sync + 'static,
+    P: Send + Sync + 'static,
+    PE: PathExtractor + Send + Sync + 'static,
+{
+    /// Binds a new `PathExtractor` to the associated routes.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// # #[macro_use]
+    /// # extern crate log;
+    /// # extern crate gotham;
+    /// # #[macro_use]
+    /// # extern crate gotham_derive;
+    /// # extern crate hyper;
+    /// #
+    /// # use hyper::Response;
+    /// # use gotham::router::Router;
+    /// # use gotham::router::builder::*;
+    /// # use gotham::state::State;
+    /// #
+    /// fn handler(_state: State) -> (State, Response) {
+    ///     // Implementation elided.
+    /// #   unimplemented!()
+    /// }
+    ///
+    /// #[derive(StateData, PathExtractor, StaticResponseExtender)]
+    /// struct MyPathExtractor {
+    /// #   #[allow(dead_code)]
+    ///     id: u32,
+    /// }
+    ///
+    /// #
+    /// # fn router() -> Router {
+    /// build_simple_router(|route| {
+    ///     route.associate("/resource/:id", |assoc| {
+    ///         let mut assoc = assoc.with_path_extractor::<MyPathExtractor>();
+    ///         assoc.get().to(handler);
+    ///     });
+    /// })
+    /// # }
+    /// # fn main() { router(); }
+    /// ```
+    pub fn with_path_extractor<'b, NPE>(&'b mut self) -> AssociatedRouteBuilder<'b, C, P, NPE>
```

Comment:
Should this also have a `with_query_extractor`? If so I am wondering if a Trait that could be applied here and to `DefineSingleRoute` may be useful if we want to support extra extractors in the future (e.g. body extractors, say x-www-form-urlencoded).

Reply:
Arguably yes, arguably no. I'm tending towards "no" because associated routes are bound to a _path_, and so the _path_ extractor makes sense to specify at the top level. Though as I say that, I imagine that a ticket will show up in a few months suggesting that its omission was a silly oversight.

My rule of thumb with factoring things into traits in the router builder is that it can't make the API documentation **harder** to read, so I'm not opposed as long as it fits the bill there. If a little duplication makes the type soup more palatable, I'm tending to err on that side.

Reply:
No worries re: type soup/Traits and readability, reasonable trade offs.

wrt `with_query_extractor` I see your point the `Struct` being extracted into, if we leave it at the `route` level, could certainly just employ `Option` fields to make it suitable for each associated route, which is ok but not without draw backs.

I did consider however that this is something that could be achieved if you simply wrote out multiple `route.verb("/path").with_query_extractor...`. For that reason alone I felt it made sense to have it here when I was looking this over.

I could come across to the "no" argument but flexibility in the router builder, where it is reasonable API/code wise, seems like a good "goal".

Reply:
Doesn't hurt to have it but also have it rarely used. I'll add it.